zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/SpotRestorationTool.java

```diff
@@ -0,0 +1,276 @@
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Scanner;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Lists;
+import org.apache.jute.Record;
+import org.apache.zookeeper.CreateMode;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.ZooDefs;
+import org.apache.zookeeper.ZooKeeper;
+import org.apache.zookeeper.data.ACL;
+import org.apache.zookeeper.data.Stat;
+import org.apache.zookeeper.server.DataNode;
+import org.apache.zookeeper.server.DataTree;
+import org.apache.zookeeper.server.persistence.FileTxnSnapLog;
+import org.apache.zookeeper.txn.TxnDigest;
+import org.apache.zookeeper.txn.TxnHeader;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This is a tool for automatic restoration of ZK data based on a zk node path using a collection of
+ * snapshot and transaction logs.
+ */
+public class SpotRestorationTool {
+  private static final Logger LOG = LoggerFactory.getLogger(SpotRestorationTool.class);
+
+  File dataDir;
+  ZooKeeper zk;
+  FileTxnSnapLog snapLog;
+  boolean restoreRecursively;
+  String targetZNodePath;
+  List<String> messages;
+
+  /**
+   * Constructor
+   * @param dataDir The directory contains snapshot and transaction log files to be used for restoration
+   * @param zk A client connection to the zk server to be restored
+   * @param targetZNodePath The znode path to start the restoration
+   * @param restoreRecursively If true then restore the whole sub-data tree of the target znode;
+   *                           if false then restore the target znode only
+   * @throws IOException
+   */
+  public SpotRestorationTool(File dataDir, ZooKeeper zk, String targetZNodePath,
+      boolean restoreRecursively) throws IOException {
+    if (dataDir == null || !dataDir.exists()) {
+      throw new IllegalArgumentException(
+          "The provided dataDir is null or does not exist, please check the input.");
+    }
+    this.dataDir = dataDir;
+    this.zk = zk;
+    this.snapLog = new FileTxnSnapLog(dataDir, dataDir);
+    this.targetZNodePath = targetZNodePath;
+    this.restoreRecursively = restoreRecursively;
+    this.messages = Lists.newArrayList();
+  }
+
+  /**
+   * Run the spot restoration.
+   * 1. If a node exists in restored data tree but not in the zk server, create the node in the server
+   * 2. If a node exists in zk server but not the restored data tree, show messages ask user to manually delete the node
+   * 3. If a node exists in both zk server and restored data tree, show message ask user to take care of the node
+   * @throws IOException
+   */
+  public void run() throws IOException {
+    DataTree dataTree = new DataTree();
+    long zxidRestored =
+        snapLog.restore(dataTree, new ConcurrentHashMap<>(), new FileTxnSnapLog.PlayBackListener() {
```

Comment:
Does this need to be a `ConcurrentHashMap`?

Reply:
This has to be a `ConcurrentHashMap` because the save method in FileTxnSnapLog has this signature:
 public void save(DataTree dataTree, ConcurrentHashMap<Long, Integer> sessionsWithTimeouts, boolean syncSnap). And a regular map cannot be cast into ConcurrentHashMap so it will produce error.